lxd/db/cluster/entities.go

```diff
@@ -81,3 +85,49 @@ func init() {
 		EntityTypes[name] = code
 	}
 }
+
+// URLToEntityType parses a raw URL string and returns the entity type, the project (if present), and the
+// path arguments. The returned project is not automatically set to "default", this is the responsibility of the caller.
+// An error is returned if the URL is not recognised.
+func URLToEntityType(rawURL string) (int, string, []string, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return -1, "", nil, err
+	}
+
+	// We need to space separate the path because fmt.Sscanf uses this as a delimiter.
+	spaceSeparatedURLPath := strings.Replace(u.Path, "/", " / ", -1)
+	for entityType, entityURI := range EntityURIs {
+		entityPath, _, _ := strings.Cut(entityURI, "?")
```

Comment:
Can we use url.Parse here for consistency in extracting the path?

Reply:
This gives a `invalid URL escape "%s" error unfortunately.

Perhaps we should be using regexes here instead.

There are now quite a few places where we need to parse urls and or mux vars into something meaningful outside of the handlers themselves. My initial plan was to utilise the fact that the mux router already does this but it meant parameterising the whole API (naming all routes and having consts for mux vars) but this led to a lot of potentially unnecessary changes and also didn't solve the case for parsing a url contained in an operation resource. In the end, this method seemed a reasonably happy middle ground.

Reply:
Yes this seems fine. Using regexes would be more complicated in my view.